### ISSUE TITLE
Compacting graph container with multiple nodes broken

### ISSUE THREAD

**Issue by gkellogg** [w3c/json-ld-api#143]
As described in https://github.com/w3c/json-ld-framing/issues/27#issuecomment-526768933, there is an issue when compacting named graphs with multiple top-level nodes under a property with `@container: @graph`.

Consider the following:

```json
{
  "@context": {
    "@version": 1.1,
    "@vocab": "http://example.org/"
  },
  "input": {
    "@graph": [
      {"value": "x"},
      {"value": "y"}
    ]
  }
}
```
When compacted using:
```json
{
  "@context": {
    "@version": 1.1,
    "@vocab": "http://example.org/",
    "input": {"@container": "@graph"}
  }
}
```
Generates the following:
```json
{
  "@context": {
    "@version": 1.1,
    "@vocab": "http://example.org/",
    "input": {"@container": "@graph"}
  },
  "input": [
    {"value": "x"},
    {"value": "y"}
  ]
}
```

But this describes two different named graphs.

We should either not select the term for compaction if there is more than one top-level node, or use either `@graph` or `@included` to wrap them all together, which pretty much undoes the affect of `@container: @graph`.

**Comment by gkellogg** [comment]
I think we should solve this by not selecting "input", as it can't express the result properly. Adding the use of `@graph` or `@included` would be just as if we weren't using a graph container, and could end up allocating a blank node for a graph name which isn't actually used.

**Comment by dlongley** [comment]
@gkellogg,

Hmm, or is it the case that we have an expansion bug here? Shouldn't the above input actually be equivalent to this (note `@graph: @container` is removed)?:

```js
{
  "@context": {
    "@version": 1.1,
    "@vocab": "http://example.org/"
  },
  "input": {
    "@graph": [{
      "@graph": [
        {"value": "x"},
        {"value": "y"}
      ]
    }]
  }
}
```

When using `@graph` inside of an `@graph: @container` it would seem to me to generate yet another blank node named graph. Which would mean that we'd want the expanded output to be:

```js
[{
  "http://example.org/input": [{
    "@graph": [{
      "@graph": [{
        "http://example.org/value": [{"@value": "x"}]
      }, {
        "http://example.org/value": [{"@value": "y"}]
      }]
    }]
  }]
}]
```

But, instead, it's dropping an `@graph`:

```js
[{
  "http://example.org/input": [{
    "@graph": [{
      "http://example.org/value": [{"@value": "x"}]
    }, {
      "http://example.org/value": [{"@value": "y"}]
    }]
  }]
}]
```

**Comment by dlongley** [comment]
This step of the [Expansion Algorithm](https://w3c.github.io/json-ld-api/#algorithm-2) has a caveat that causes the above to happen:

> 13.8.3.6.1) If container mapping includes `@graph` and ***if item is not a graph object***, set item to a new map containing the key-value pair `@graph-item`, ensuring that the value is represented using an array.

That step is in there to prevent double expanding graphs -- but I think that might just be wrong. I'm going through the tests now to see what's going on there, and so far all of the expansion tests are just testing to make sure we don't do this double expansion whenever a `@container: @graph` is used. But, it seems to me, we would *want* it to do the double expansion.

**Comment by dlongley** [comment]
Yes, that seems to be all that's tested for, to handle this special case. There are toRDF tests as well but those are just repeats of these special case expansion tests. It seems like they should either be changed or removed if we can agree that dropping an extra `@graph` into the value *should* change what's being expressed. I think it should and I'm not sure why it was ever any other way (perhaps it was something wacky I did in the original implementation). Ideally we'd just drop this special handling.

**Comment by gkellogg** [comment]
> @gkellogg,
> 
> Hmm, or is it the case that we have an expansion bug here? Shouldn't the above input actually be equivalent to this (note `@graph: @container` is removed)?:
> 
> ```js
> {
>   "@context": {
>     "@version": 1.1,
>     "@vocab": "http://example.org/"
>   },
>   "input": {
>     "@graph": [{
>       "@graph": [
>         {"value": "x"},
>         {"value": "y"}
>       ]
>     }]
>   }
> }
> ```
> 
> When using `@graph` inside of an `@graph: @container` it would seem to me to generate yet another blank node named graph. Which would mean that we'd want the expanded output to be:

That's not what `@graph` with multiple values means, from [4.9 Named Graphs](https://w3c.github.io/json-ld-syntax/#named-graphs), when not used at the top-level, `@graph` introduces a single named graph, taking the graph name from the node containing the `@graph`, if present, or a blank node otherwise. So, the double `@graph` isn't necessary (and would be extremely confusing).

Your example would yield the following quads:
```ntriples
_:b0 <http://example.org/input> _:b1 .
_:b4 <http://example.org/value> "y" _:b2 .
_:b3 <http://example.org/value> "x" _:b2 .
```
Without the extra `@graph`, you get the following:
```ntriples
_:b0 <http://example.org/input> _:b1 .
_:b3 <http://example.org/value> "y" _:b1 .
_:b2 <http://example.org/value> "x" _:b1 .
```
which is, what we want.

Going back to JSON-LD, my distiller gives the following:
```json
{
  "@graph": [{
    "@id": "_:b0",
    "http://example.org/input": {"@id": "_:b1"}
  },
  {
    "@id": "_:b1",
    "@graph": [{
      "@id": "_:b3",
      "http://example.org/value": "y"
    }, {
      "@id": "_:b2",
      "http://example.org/value": "x"
    }
    ]
  }]
}
```
Which could then be trivially framed back to:
```json
{
  "http://example.org/input": {
    "@graph": [{
      "http://example.org/value": "y"
    },{
      "http://example.org/value": "x"
    }]
  }
}
```

**Comment by dlongley** [comment]
> So, the double @graph isn't necessary (and would be extremely confusing).

I don't think this is about it being necessary, rather what it means if someone does it.

A term with `@container: @graph` means that its values will be interpreted as being in an implicit blank node named graph. If someone adds an additional `@graph` in there -- I don't think we should be doing some extra lifting to remove it for them, rather, we should assume that it is what they meant to express. Doing otherwise, I think, is more confusing. So if an author is using such a term, don't add `@graph` to a value unless you really want to create yet *another* named graph. And, yes, the quads would be just as you say.

Imagine a scenario where you wanted to make statements about another non-blank node named graph in a graph container. How would you do it? Imagine someone's attempt:

```js
{
  "@context": {
    "@version": 1.1,
    "@vocab": "http://example.org/",
    "input": {"@container": "@graph"}
  },
  "input": {
    "@id": "ex:someIdentifier",
    "ex:about": "this graph",
    "@graph": [
        {"value": "x"},
        {"value": "y"}
      ]
    }
  }
}
```

If you converted to quads, you'd get:

```
<ex:someIdentifier> <ex:about> "this graph" .
_:b0 <http://example.org/input> <ex:someIdentifier> .
_:b1 <http://example.org/value> "x" <ex:someIdentifier> .
_:b2 <http://example.org/value> "y" <ex:someIdentifier> .
```

But this is wrong! It's as if there was no `@container: @graph` at all. That "ex:about" statement is supposed to be wrapped up and isolated in a blank node named graph, but it's bled into the default graph instead. The quads you *should* be getting would be more like:

```
<ex:someIdentifier> <ex:about> "this graph" _:b1 .
_:b0 <http://example.org/input> _:b1 .
_:b2 <http://example.org/value> "x" <ex:someIdentifier> .
_:b3 <http://example.org/value> "y" <ex:someIdentifier> .
```

I'm asserting that this:

```js
{
  "@context": {
    "@version": 1.1,
    "@vocab": "http://example.org/",
    "input": {"@container": "@graph"}
  },
  "input": {
    "@id": "ex:someIdentifier",
    "ex:about": "this graph",
    "@graph": [
        {"value": "x"},
        {"value": "y"}
      ]
    }
  }
}
```

And this:

```js
{
  "@context": {
    "@version": 1.1,
    "@vocab": "http://example.org/"
  },
  "input": {
    "@id": "ex:someIdentifier",
    "ex:about": "this graph",
    "@graph": [
        {"value": "x"},
        {"value": "y"}
      ]
    }
  }
}
```

Are not the same thing. Why should we treat them as such?

**Comment by gkellogg** [comment]
> > So, the double @graph isn't necessary (and would be extremely confusing).
> 
> I don't think this is about it being necessary, rather what it means if someone does it.
> 
> A term with `@container: @graph` means that its values will be interpreted as being in an implicit blank node named graph. If someone adds an additional `@graph` in there -- I don't think we should be doing some extra lifting to remove it for them, rather, we should assume that it is what they meant to express. Doing otherwise, I think, is more confusing. So if an author is using such a term, don't add `@graph` to a value unless you really want to create yet _another_ named graph. And, yes, the quads would be just as you say.
> 
> Imagine a scenario where you wanted to make statements about another non-blank node named graph in a graph container. How would you do it? Imagine someone's attempt:
> 
> ```js
> {
>   "@context": {
>     "@version": 1.1,
>     "@vocab": "http://example.org/",
>     "input": {"@container": "@graph"}
>   },
>   "input": {
>     "@id": "ex:someIdentifier",
>     "ex:about": "this graph",
>     "@graph": [
>       {"value": "x"},
>       {"value": "y"}
>     ]
>   }
> }
> ```
> 
> If you converted to quads, you'd get:
> 
> ```
> <ex:someIdentifier> <ex:about> "this graph" .
> _:b0 <http://example.org/input> <ex:someIdentifier> .
> _:b1 <http://example.org/value> "x" <ex:someIdentifier> .
> _:b2 <http://example.org/value> "y" <ex:someIdentifier> .
> ```
> 
> But this is wrong! It's as if there was no `@container: @graph` at all. That "ex:about" statement is supposed to be wrapped up and isolated in a blank node named graph, but it's bled into the default graph instead. The quads you _should_ be getting would be more like:

It is wrong, and that's not what I get on my [distiller](http://rdf.greggkellogg.net/distiller?command=serialize&format=jsonld&output_format=nquads).

> ```
> <ex:someIdentifier> <ex:about> "this graph" _:b1 .
> _:b0 <http://example.org/input> _:b1 .
> _:b2 <http://example.org/value> "x" <ex:someIdentifier> .
> _:b3 <http://example.org/value> "y" <ex:someIdentifier> .
> ```

That _is_ what I get. That's also what I get on the [json-ld.org playground](https://json-ld.org/playground/#startTab=tab-nquads&json-ld=%7B%22%40context%22%3A%7B%22%40version%22%3A1.1%2C%22%40vocab%22%3A%22http%3A%2F%2Fexample.org%2F%22%2C%22input%22%3A%7B%22%40container%22%3A%22%40graph%22%7D%7D%2C%22input%22%3A%7B%22%40id%22%3A%22ex%3AsomeIdentifier%22%2C%22ex%3Aabout%22%3A%22this%20graph%22%2C%22%40graph%22%3A%5B%7B%22value%22%3A%22x%22%7D%2C%7B%22value%22%3A%22y%22%7D%5D%7D%7D).
 
> I'm asserting that this:
> 
> ```js
> {
>   "@context": {
>     "@version": 1.1,
>     "@vocab": "http://example.org/",
>     "input": {"@container": "@graph"}
>   },
>   "input": {
>     "@id": "ex:someIdentifier",
>     "ex:about": "this graph",
>     "@graph": [
>       {"value": "x"},
>       {"value": "y"}
>     ]
>   }
> }
> ```
> 
> And this:
> 
> ```js
> {
>   "@context": {
>     "@version": 1.1,
>     "@vocab": "http://example.org/"
>   },
>   "input": {
>     "@id": "ex:someIdentifier",
>     "ex:about": "this graph",
>     "@graph": [
>       {"value": "x"},
>       {"value": "y"}
>     ]
>   }
> }
> ```
> 
> Are not the same thing. Why should we treat them as such?

We don't.

**Comment by dlongley** [comment]
I was generating that output using the playground. I'm not sure what happened, but it was probably user (me) error.

So here's where things get weird then:

This:
```js
{
  "@context": {
    "@version": 1.1,
    "@vocab": "http://example.org/",
    "input": {"@container": "@graph"}
  },
  "input": {
    "ex:about": "this graph",
    "@graph": [
      {"value": "x"},
      {"value": "y"}
    ]
  }
}
```

Yields these quads (if I'm not misusing the playground again somehow):
```
_:b0 <http://example.org/input> _:b1 .
_:b2 <ex:about> "this graph" _:b1 .
_:b3 <http://example.org/value> "x" _:b2 .
_:b4 <http://example.org/value> "y" _:b2 .
```

But this:
```js
{
  "@context": {
    "@version": 1.1,
    "@vocab": "http://example.org/",
    "input": {"@container": "@graph"}
  },
  "input": {
    "@graph": [
      {"value": "x"},
      {"value": "y"}
    ]
  }
}
```

Yields these quads, eliminating the extra graph:
```
_:b0 <http://example.org/input> _:b1 .
_:b2 <http://example.org/value> "x" _:b1 .
_:b3 <http://example.org/value> "y" _:b1 .
```

Why would we want this? What benefit is there? If someone puts that extra `@graph` in there, I think we should presume it's what they meant to do. If we're going to have a special case in the expansion algorithm it should be justified -- do we know why we have it?

**Comment by dlongley** [comment]
So I'm not sure what's going on with me or the playground, but I just tried this:

```js
{
  "@context": {
    "@version": 1.1,
    "@vocab": "http://example.org/",
    "input": {"@container": "@graph"}
  },
  "input": [{
    "@id": "ex:someIdentifier",
    "@graph": [
      {"value": "x"},
      {"value": "y"}
    ]
  }]
}
```

And that produced these quads:

```
_:b0 <http://example.org/input> <ex:someIdentifier> .
_:b1 <http://example.org/value> "x" <ex:someIdentifier> .
_:b2 <http://example.org/value> "y" <ex:someIdentifier> .
```

There's no blank node named graph in there, which is wrong. Which means it's being treated the same as:

```js
```js
{
  "@context": {
    "@version": 1.1,
    "@vocab": "http://example.org/"
  },
  "input": [{
    "@id": "ex:someIdentifier",
    "@graph": [
      {"value": "x"},
      {"value": "y"}
    ]
  }]
}
```

It should instead produce:

```
_:b0 <http://example.org/input> _:b1 .
_:b2 <http://example.org/value> "x" <ex:someIdentifier> .
_:b3 <http://example.org/value> "y" <ex:someIdentifier> .
```

Am I misusing the playground again?

**Comment by gkellogg** [comment]
> I was generating that output using the playground. I'm not sure what happened, but it was probably user (me) error.
> 
> So here's where things get weird then:
> 
> This:
> 
> ```js
> {
>   "@context": {
>     "@version": 1.1,
>     "@vocab": "http://example.org/",
>     "input": {"@container": "@graph"}
>   },
>   "input": {
>     "ex:about": "this graph",
>     "@graph": [
>       {"value": "x"},
>       {"value": "y"}
>     ]
>   }
> }
> ```
> 
> Yields these quads (if I'm not misusing the playground again somehow):
> 
> ```
> _:b0 <http://example.org/input> _:b1 .
> _:b2 <ex:about> "this graph" _:b1 .
> _:b3 <http://example.org/value> "x" _:b2 .
> _:b4 <http://example.org/value> "y" _:b2 .
> ```
> 
> But this:
> 
> ```js
> {
>   "@context": {
>     "@version": 1.1,
>     "@vocab": "http://example.org/",
>     "input": {"@container": "@graph"}
>   },
>   "input": {
>     "@graph": [
>       {"value": "x"},
>       {"value": "y"}
>     ]
>   }
> }
> ```
> 
> Yields these quads, eliminating the extra graph:
> 
> ```
> _:b0 <http://example.org/input> _:b1 .
> _:b2 <http://example.org/value> "x" _:b1 .
> _:b3 <http://example.org/value> "y" _:b1 .
> ```
> 
> Why would we want this? What benefit is there? If someone puts that extra `@graph` in there, I think we should presume it's what they meant to do. If we're going to have a special case in the expansion algorithm it should be justified -- do we know why we have it?

Looks like some bug in generating the node map, as the `@graph` should create a new graph name, and not `@default`. Perhaps in step 6.10, where `id` doesn't get set? I'll create a test for this.

**Comment by gkellogg** [comment]
> ```js
> {
>   "@context": {
>     "@version": 1.1,
>     "@vocab": "http://example.org/",
>     "input": {"@container": "@graph"}
>   },
>   "input": [{
>     "@id": "ex:someIdentifier",
>     "@graph": [
>       {"value": "x"},
>       {"value": "y"}
>     ]
>   }]
> }
> ```

Must be the same node map bug, as my implementation does just the same thing.

**Comment by gkellogg** [comment]
I created #144 for these issues.

**Comment by dlongley** [comment]
@gkellogg, I believe that if the special case code I mentioned in https://github.com/w3c/json-ld-api/issues/143#issuecomment-526858659 is removed this problem goes away.

**Comment by dlongley** [comment]
It seems to me that the purpose of this special case code may have been to allow `@graph` to be used like it is at the root level, to specify multiple roots in the `@default` graph, except here, for a blank node named graph.

If that's the case, it seems to me that trying to enable `@graph` to work that way in this one particular case is not preferable to using `@included` now that we have it. I believe it's also leading to these other bugs and/or points of confusion.

**Comment by gkellogg** [comment]
> @gkellogg,
> 
> Hmm, or is it the case that we have an expansion bug here? Shouldn't the above input actually be equivalent to this (note `@graph: @container` is removed)?:
> 
> ```js
> {
>   "@context": {
>     "@version": 1.1,
>     "@vocab": "http://example.org/"
>   },
>   "input": {
>     "@graph": [{
>       "@graph": [
>         {"value": "x"},
>         {"value": "y"}
>       ]
>     }]
>   }
> }
> ```
I made an error (which I've corrected), as I was trying to illustrate the compaction problem.

**Comment by gkellogg** [comment]
> This step of the [Expansion Algorithm](https://w3c.github.io/json-ld-api/#algorithm-2) has a caveat that causes the above to happen:
> 
> > 13.8.3.6.1) If container mapping includes `@graph` and _**if item is not a graph object**_, set item to a new map containing the key-value pair `@graph-item`, ensuring that the value is represented using an array.
> 
> That step is in there to prevent double expanding graphs -- but I think that might just be wrong. I'm going through the tests now to see what's going on there, and so far all of the expansion tests are just testing to make sure we don't do this double expansion whenever a `@container: @graph` is used. But, it seems to me, we would _want_ it to do the double expansion.

I can't think of a use case for double graphs. In other cases, if we have a container setting where the value already. Consider the following:

```json
{
  "@context": {"@vocab": "http://example/", "list": {"@container": "@list"}},
  "list": {"@list": [1,2]}
}
```
This does not create a double list.

Also, `@container: @id`, or `@container: `@type` work only when the value is a map, so it's entirely in keeping with this for the value of a property with `@container: @graph` to not do a double expansion of graphs.

**Comment by gkellogg** [comment]
> @gkellogg, I believe that if the special case code I mentioned in [#143 (comment)](https://github.com/w3c/json-ld-api/issues/143#issuecomment-526858659) is removed this problem goes away.

Given the potential to mix simple graphs with more complicated ones, it probably is, indeed, best to remove that special case.

**Comment by dlongley** [comment]
> I can't think of a use case for double graphs.

Me either; I don't really think it's about supporting them vs. removing special case behavior we don't need and that seems like it's gumming up the works.

> Given the potential to mix simple graphs with more complicated ones, it probably is, indeed, best to remove that special case.

:+1:

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-09-06-json-ld#section4-1).

- `RESOLVED:  Close api #143 as resolved by api PR #145 and #146`
<details><summary><i class="differentiate">View the transcript</i></summary>
Framing blank nodes<br/>
Rob Sanderson: last discussion we agreed that we couldn’t solve it on a call<br/>
… so gkellog and dlongley went off to look at it<br/>
Gregg Kellogg: we found a problem in a framing test where <code>@container</code> : <code>@graph</code> got mangled in re-expansion<br/>
… a bug in the compaction algo<br/>
… if the value is an array, it puts them in an <code>@included</code> block<br/>
… i tried [s solution] but it turned out not to be defined well enough<br/>
Rob Sanderson: all of that is solved and merged?<br/>
Gregg Kellogg: yep<br/>
Gregg Kellogg: See API PR #146<br/>
Gregg Kellogg: See API PR #145<br/>
Proposed resolution: Close framing #27 as not being the issue, and the real issues being addressed is api #143, solved by api PRs # 145 and #146 (Rob Sanderson)<br/>
Rob Sanderson: +1<br/>
Benjamin Young: +1<br/>
Dave Longley: +1<br/>
Gregg Kellogg: +1<br/>
Ivan Herman: +1<br/>
Adam Soroka: +1<br/>
Pierre-Antoine Champin: +1<br/>
Ruben Taelman: +1<br/>
Rob Sanderson: RESOLVE: Close framing #27 as not being the issue, and the real issues being addressed is api #143, solved by api PRs # 145 and #146<br/>
Proposed resolution: Close api #143 as resolved by api PR #145 and #146 (Rob Sanderson)<br/>
Rob Sanderson: +1<br/>
Ivan Herman: +1<br/>
Ruben Taelman: +1<br/>
Gregg Kellogg: +1<br/>
Adam Soroka: +1<br/>
Pierre-Antoine Champin: +1<br/>
Benjamin Young: +1<br/>
Dave Longley: +1<br/>
Resolution #2: Close api #143 as resolved by api PR #145 and #146<br/>
</details>
